crates/sui-types/src/crypto.rs

```diff
@@ -262,77 +266,207 @@ impl std::fmt::Debug for Signature {
     }
 }
 
-impl Signature {
-    pub fn new<T>(value: &T, secret: &dyn signature::Signer<Signature>) -> Self
-    where
-        T: Signable<Vec<u8>>,
-    {
-        let mut message = Vec::new();
-        value.write(&mut message);
-        secret.sign(&message)
+//
+// Ed25519 Sui Signature port
+//
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct Ed25519SuiSignature(
+    #[schemars(with = "Base64")]
+    #[serde_as(as = "Readable<Base64, Bytes>")]
+    [u8; Ed25519PublicKey::LENGTH + Ed25519Signature::LENGTH + FLAG_LENGTH],
+);
+
+impl SuiSignatureInner for Ed25519SuiSignature {
+    type Sig = Ed25519Signature;
+    type PubKey = Ed25519PublicKey;
+    type KeyPair = Ed25519KeyPair;
+    const LENGTH: usize = Ed25519PublicKey::LENGTH + Ed25519Signature::LENGTH + FLAG_LENGTH;
+    const FLAG: [u8; FLAG_LENGTH] = [0xed];
+}
+
+impl AsRef<[u8]> for Ed25519SuiSignature {
+    fn as_ref(&self) -> &[u8] {
+        self.0.as_ref()
     }
+}
 
-    pub fn new_empty() -> Self {
-        Self([0u8; SUI_SIGNATURE_LENGTH])
+impl signature::Signature for Ed25519SuiSignature {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        if bytes.len() != Self::LENGTH {
+            return Err(signature::Error::new());
+        }
+        let mut sig_bytes = [0; Self::LENGTH];
+        sig_bytes.copy_from_slice(bytes);
+        Ok(Self(sig_bytes))
     }
+}
+
+// impl signature::Signer<Signature> for Ed25519KeyPair {
+//     fn try_sign(&self, msg: &[u8]) -> Result<Signature, signature::Error> {
+//         Ok(Ed25519SuiSignature::new(self, msg)
+//             .map_err(|_| signature::Error::new())?.into())
+//     }
+// }
 
-    pub fn signature_bytes(&self) -> &[u8] {
-        &self.0[..AccountSignature::LENGTH]
+//
+// Ed25519 Sui Signature port
+//
+#[serde_as]
+#[derive(Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct Secp256k1SuiSignature(
+    #[schemars(with = "Base64")]
+    #[serde_as(as = "Readable<Base64, Bytes>")]
+    [u8; Secp256k1PublicKey::LENGTH + Secp256k1Signature::LENGTH + FLAG_LENGTH],
+);
+
+impl SuiSignatureInner for Secp256k1SuiSignature {
```

Comment:
The work done on the narwhal side was super useful - we barely have to reimplement anything.